nextcord/http.py

```diff
@@ -151,48 +476,172 @@ def __exit__(
+
+        self._global_rate_limits[auth] = rate_limit
+        return rate_limit
+
+    def _make_url_rate_limit(self, method: str, route: Route, auth: Optional[str]) -> RateLimit:
+        _log.debug(
+            "Making URL rate limit for %s %s %s", method, route.bucket, _get_logging_auth(auth)
+        )
+        ret = RateLimit(
+            time_offset=self._time_offset, use_reset_timestamp=self._ratelimit_use_timestamp
+        )
+        self._url_rate_limits[(method, route.bucket, auth)] = ret
+        return ret
+
+    def _set_url_rate_limit(
+        self, method: str, route: Route, auth: Optional[str], rate_limit: RateLimit
+    ) -> None:
+        self._url_rate_limits[(method, route.bucket, auth)] = rate_limit
+
+    def _get_url_rate_limit(
+        self, method: str, route: Route, auth: Optional[str]
+    ) -> Optional[RateLimit]:
+        return self._url_rate_limits.get((method, route.bucket, auth), None)
+
+    def set_default_auth(self, auth: Optional[str]) -> None:
+        self._default_auth = auth
+
+    def _make_headers(
+        self,
+        original_headers: dict[str, str],
+        *,
+        auth: Optional[str] = MISSING,
+    ) -> dict[str, str]:
+        """Creates a new dictionary of headers, without overwriting values from the given headers.
+
+        Parameters
+        ----------
+        original_headers: :class:`dict`[:class:`str`, :class:`str`]
+            Headers to make a shallow copy of.
+        auth: :class:`str` | `None` | `MISSING`
+            Authorization string to use. Will not auto-format given tokens. For example, a bot token must be provided
+            as "Bot <bot token>". If set to `None`, no authorization header will be added. If left unset, the default
+            auth string will be used. (if the default is not set, no auth will be used.)
+
+        Returns
+        -------
+        :class:`dict`[:class:`str`, :class:`str`]
+            Modified headers to use.
+        """
+        ret = original_headers.copy()
+
+        if "Authorization" not in ret:
+            if auth is None:
+                pass  # We do nothing, this is here to make the logic easier to read.
+            elif auth is MISSING:
+                if self._default_auth is not None:
+                    ret["Authorization"] = self._default_auth
+            else:  # auth isn't None or MISSING, so it must be something.
+                ret["Authorization"] = auth
+
+        if "User-Agent" not in ret and self._user_agent:
+            ret["User-Agent"] = self._user_agent
+
+        return ret
 
     def recreate(self) -> None:
         if self.__session.closed:
             self.__session = aiohttp.ClientSession(
-                connector=self.connector,
+                connector=self._connector,
                 ws_response_class=DiscordClientWebSocketResponse,
-                headers={
-                    "User-Agent": _USER_AGENT,
-                },
             )
```

Comment:
Is this safe? (I know it's not your code)

Reply:
It's not, it partially reverts #1114 (and thus #1002 becomes an issue again).

Reply:
Currently I have the user agent stuff handled in `HTTPClient._make_headers`, but I can add it back into the `ClientSession` itself if desired.

Reply:
I was more concerned about the creation of a ClientSession potentionally without a running loop / without an async context. (bot crashed, user calls recreate with a closed loop, ...)

Reply:
Fixing, and making `Client.clear()` an async function as well.